src/metrics/rules/active_ruleset.go

```diff
@@ -45,6 +48,69 @@ type activeRuleSet struct {
 	tagsFilterOpts  filters.TagsFilterOptions
 	newRollupIDFn   metricid.NewIDFn
 	isRollupIDFn    metricid.MatchIDFn
+
+	useFastMatch bool
+	// The key is the value of the __name__ label, the value is the mappingRule
+	mappingRuleIndex map[string][]*mappingRule
+	// The key is the value of the __name__ label, the value is rollupRule
+	rollupRuleIndex map[string][]*rollupRule
+}
+
+func newActiveRuleSetFast(
+	version int,
+	mappingRules []*mappingRule,
+	rollupRules []*rollupRule,
+	tagsFilterOpts filters.TagsFilterOptions,
+	newRollupIDFn metricid.NewIDFn,
+	isRollupIDFn metricid.MatchIDFn,
+) *activeRuleSet {
+	rset := newActiveRuleSet(version, mappingRules, rollupRules, tagsFilterOpts, newRollupIDFn, isRollupIDFn)
+	rset.useFastMatch = true
+
+	nowNanos := time.Now().UnixNano()
+
+	getNameTagValue := func(nameFilterValue *filters.FilterValue, snapshotString func() string) string {
+		if nameFilterValue == nil {
+			panic(fmt.Sprintf("A rule has no fitler for metric name tag: %s", snapshotString()))
+		}
+		if nameFilterValue.Negate {
+			panic(fmt.Sprintf("A rule has negated filter for metric name tag: %s", snapshotString()))
+		}
+		if strings.Contains(nameFilterValue.Pattern, "*") {
+			panic(fmt.Sprintf("A rule has wildcard filter for metric name tag %s", snapshotString()))
+		}
+		return nameFilterValue.Pattern
+	}
+
+	rset.mappingRuleIndex = make(map[string][]*mappingRule)
+	for _, rule := range mappingRules {
+		snapshot := rule.activeSnapshot(nowNanos)
+		if snapshot == nil {
+			panic(fmt.Sprintf("mapping rule has no active snapshot: %+v", rule))
+		}
+		nameTagValue := getNameTagValue(snapshot.filter.NameFilterValue(), func() string { return fmt.Sprintf("%+v", snapshot) })
+		rset.mappingRuleIndex[nameTagValue] = append(rset.mappingRuleIndex[nameTagValue], rule)
+	}
+	rset.rollupRuleIndex = make(map[string][]*rollupRule)
+	for _, rule := range rollupRules {
+		snapshot := rule.activeSnapshot(nowNanos)
+		if snapshot == nil {
+			panic(fmt.Sprintf("rollup rule has no active snapshot: %+v", rule))
+		}
+		nameTagValue := getNameTagValue(snapshot.filter.NameFilterValue(), func() string { return fmt.Sprintf("%+v", snapshot) })
+		rset.rollupRuleIndex[nameTagValue] = append(rset.rollupRuleIndex[nameTagValue], rule)
+	}
+	// TODO: use a proper logger
+	logger := log.Default()
```

Comment:
nit: use `zap.Logger` from opts, see `promremote/storage.go`

Reply:
Things are a bit different here. promStorage got an already initialized logger from its call site. Here, the call site doesn't pass down a logger. That's why the entire file doesn't log anything before this PR.
<img width="946" alt="image" src="https://github.com/user-attachments/assets/624833a2-4844-4b0e-901e-5694502b3572">